fenicsadapter/adapter_core.py

```diff
@@ -81,39 +83,42 @@ def filter_point_sources(point_sources, filter_out):
     return filtered_point_sources
 
 
-def convert_fenics_to_precice(data, sample_points):
+def convert_fenics_to_precice(function, coupling_subdomain, fenics_dimensions, dimensions):
     """
     Converts data of type dolfin.Function into Numpy array for all x and y coordinates on the boundary.
 
     Parameters
     ----------
-    data : FEniCS function
+    function : FEniCS function
         A FEniCS function referring to a physical variable in the problem.
-    sample_points : array_like
-        The coordinates of the vertices in a numpy array [N x D] where
-        N = number of vertices and D = dimensions of geometry.
 
     Returns
     -------
     array : array_like
         Array of FEniCS function values at each point on the boundary.
     """
-    if type(data) is dolfin.Function:
-        x_all, y_all = sample_points[:, 0], sample_points[:, 1]
-        return np.array([data(x, y) for x, y in zip(x_all, y_all)])
+    _, _, _, _, local_ids, _ = get_coupling_boundary_vertices(function.function_space().mesh(), function.function_space(),
+                                                              coupling_subdomain, fenics_dimensions, dimensions)
+    if type(function) is dolfin.Function:
+        func_vector = function.vector().get_local()
+        print("N dof coordinates = {}".format(len(function.function_space().tabulate_dof_coordinates())))
+        print("len func_vector = {}".format(len(func_vector)))
+        return np.array([func_vector[lid] for lid in local_ids])
+
     else:
-        raise Exception("Cannot handle data type %s" % type(data))
+        raise Exception("Cannot handle data type {}".format(type(function)))
```

Comment:
I would suggest to move this `Exception` to the top of the function:

```
if type(function) is not dolfin.Function:
    raise Exception("Cannot handle data type {}".format(type(function)))
```

This saves us some indentation in the remainder of the function.